snapcraft/internal/lifecycle.py

```diff
@@ -112,31 +112,37 @@ def __init__(self, config, project_options):
         self.config = config
         self.project_options = project_options
         self.parts_config = config.parts
+        self._steps_run = {p: set() for p in self.config.part_names}
 
     def run(self, step, part_names=None, recursed=False):
         if part_names:
             self.parts_config.validate(part_names)
-            parts = {p for p in self.config.all_parts if p.name in part_names}
+            parts = [p for p in self.config.all_parts if p.name in part_names]
         else:
             parts = self.config.all_parts
             part_names = self.config.part_names
 
-        dirty = {p.name for p in parts if p.should_step_run('stage')}
+        run_stage = {p for p in parts if p.should_step_run('stage')}
```

Comment:
Note how `dirty` was a set of names, but in this PR it's now a set of parts. It's used to filter down the prereqs in `_run_step()` (line 144 in this PR), which is also names. Will this continue to work now that it's not a set of names?